Add Configuration Builder (and tests)

jkmassel opened the pull request:
## What?
Android uses a configuration builder which is much cleaner around mutability. This PR adds a configuration builder to iOS as well.

## Why?
The configuration has concerning mutability – it should never change after it's created because it's injected at the start. Once it starts being consumed, changes to it don't apply elsewhere, so by making it immutable it's harder to use it wrong.

## How?
Builder pattern

## Testing Instructions
Covered by new unit tests.

jkmassel committed:
Add EditorConfigurationBuilder

jkmassel committed:
Add EditorConfigurationBuilderTests

jkmassel committed:
Make Test Suite run on macOS

dcalhoun commented on Makefile:
```diff
@@ -59,4 +62,4 @@ build-swift-package: build
 	$(call XCODEBUILD_CMD, build)
 
 test-swift-package: build
-	$(call XCODEBUILD_CMD, test)
```
Will this have a negative impact on CI checks? Or does the device used not matter for the tests?

dcalhoun replied:
I reverted this in https://github.com/wordpress-mobile/GutenbergKit/pull/146/commits/26c3afa3e7a9b9db7f5bce3069d09cad3e5c2aad. It seems this was originally changed when making the test suite pass on macOS (https://github.com/wordpress-mobile/GutenbergKit/pull/146/commits/7dca3ef282d22b10253932efaee4a4dcb7552ce3), which makes sense, but I'm uncertain why we'd remove the specific device for CI runs.

dcalhoun reviewed (commented):
At a high level, this makes sense to me. Solid improvements—thank you! Testing the Demo app worked as expected. 

This is likely already understood, but building WP-iOS with these changes fails currently. Before merging this, we should open a sibling PR updating the WP-iOS implementation. 

That said, I wanted to address the following: 

> The configuration has concerning mutability – it should never change after it's created because it's injected at the start. Once it starts being consumed, changes to it don't apply elsewhere, so by making it immutable it's harder to use it wrong.

It's not entirely true that mutations do not have an impact at the moment, because we explicitly rely upon [mutating](https://github.com/wordpress-mobile/GutenbergKit/blob/f26928ac3a15ad619dd4f6ebf02402e385edce6f/ios/Sources/GutenbergKit/Sources/EditorViewController.swift#L239-L241) the configuration to set the [editor settings](https://github.com/wordpress-mobile/WordPress-iOS/blob/121f472ed898d046880d4979345fe2348f94e04e/WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift#L381-L386) and start the editor after the host app [fetches the settings](https://github.com/wordpress-mobile/WordPress-iOS/blob/121f472ed898d046880d4979345fe2348f94e04e/WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift#L352-L375). 

I am quite confident that code could/should be refactored to be less complex. There are a couple of nuances to consider: 

1. We ["warm up" the editor](https://github.com/wordpress-mobile/WordPress-iOS/blob/121f472ed898d046880d4979345fe2348f94e04e/WordPress/Classes/ViewRelated/Blog/My%20Site/MySiteViewController.swift#L173) on the _My Site_ tab with an empty configuration; 
2. [Within `init`](https://github.com/wordpress-mobile/WordPress-iOS/blob/121f472ed898d046880d4979345fe2348f94e04e/WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift#L154-L184), we initialize the editor configuration and view controller; 
3. And [within `viewDidLoad`](https://github.com/wordpress-mobile/WordPress-iOS/blob/121f472ed898d046880d4979345fe2348f94e04e/WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift#L208-L212), we display an activity indicator and fetch the editor settings. 

These nuances could likely be refactored to embrace immutable configuration while retaining the necessary loading sequence.

jkmassel commented:
@dcalhoun – I broadly agree with your comment. IMHO having two methods on the editor like `init` and `updateConfiguration` both of which take an immutable configuration would make a ton of sense, and would make it clear that simply updating the configuration object doesn't cause the underlying editor to change in any way.

I'm happy to leave this PR with you, feel free to do with it what you think is best!

dcalhoun committed:
docs: Document all editor configuration options

dcalhoun committed:
test: Expand builder tests

dcalhoun committed:
test: Fix EditorManifestTests for macOS runs

Apply iOS conditionals used in other, existing code.

dcalhoun committed:
ci: Revert Swift test script device configuration removal

This was changed while making the tests pass when running on macOS, but
it is unclear why the original CI script was changed. Presumably we
should run the tests with a fixed device on the CI.

https://github.com/wordpress-mobile/GutenbergKit/pull/146/commits/debc1c74d5d460ac59363edf5117026e1f3ffe01#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L62

dcalhoun commented:
@crazytonyli in addition to help with https://github.com/wordpress-mobile/WordPress-iOS/pull/24662#discussion_r2195212119, I welcome your review of these changes now that I've rebased it atop the latest `trunk` branch. Thank you!

dcalhoun committed:
test: Remove builder cookie assertions

This functionality was removed in the trunk branch.

jkmassel committed:
Fix cookies

dcalhoun commented:
@jkmassel unfortunately, this is a breaking change. Would you be willing to resolve the [incompatible types](https://github.com/wordpress-mobile/WordPress-iOS/pull/24662#discussion_r2195212119) and finish the WP-iOS integration? Alternatively, we could revert this change for the time being.

dcalhoun commented:
@jkmassel I'll also note that GBK cookie configuration was removed from iOS in https://github.com/wordpress-mobile/GutenbergKit/pull/157. I see that https://github.com/wordpress-mobile/GutenbergKit/pull/146/commits/fc28a0ab89d20d0e1c6c1e37e2beeb45515d73ad re-added iOS cookie configuration. 

We should either remove cookie configuration again or finish reinstating it—ensuring we [set them for the WebView](https://github.com/wordpress-mobile/GutenbergKit/commit/5ec3e9b437879b504cfed7a25ef9c60d05bf1344#diff-7d1c3b45b49105e10e9dc4ac403350f6223668df8bcd7c7d287a3eb202494a2aL40-L45) and updating the WP-iOS integration. 

Currently, WP-iOS relies upon the [shared global cookie jar](https://github.com/wordpress-mobile/WordPress-iOS/blob/248cd1c619f3289ea63ce40376f04b2004742138/WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift#L393-L413) used by the WebView. GBK's cookie logic clears that out when launching its own WebView. We'd need to update WP-iOS to address this.

dcalhoun commented:
@jkmassel it looks like this also reinstated the WebView globals removed in https://github.com/wordpress-mobile/GutenbergKit/pull/172. I suggest we remove that too. It is now unused.
